test/parallel_api/numeric/numeric.ops/inclusive_scan_by_segment.pass.cpp

```diff
@@ -53,31 +55,81 @@ struct test_inclusive_scan_by_segment
         }
     }
 
-    template <typename Iterator1, typename Iterator2, typename Size>
+    template <typename TValue>
+    struct last_accum_state
+    {
+        int idx_begin = 0;
+        int idx_end = 0;
+        TValue summ = {};
+    };
+
+    template <typename TValue, typename Iterator>
+    int
+    accum_values(Iterator host_vals, int idx_begin, int idx_end, last_accum_state<TValue>& cash)
+    {
+        if (idx_begin == cash.idx_begin && idx_end == cash.idx_end + 1 && idx_end - 1 >= idx_begin)
+            cash.summ += host_vals[idx_end - 1];
+        else
+            cash.summ = ::std::accumulate(::std::addressof(host_vals[idx_begin]), ::std::addressof(host_vals[idx_begin]) + idx_end - idx_begin, 0);
+
+        cash.idx_begin = idx_begin;
+        cash.idx_end = idx_end;
+
+        return cash.summ;
+    }
+
+#ifdef DUMP_CHECK_RESULTS
+    template <typename Iterator, typename Size>
+    void display_param(const char* msg, Iterator it, Size n)
+    {
+        std::cout << msg;
+        for (int i = 0; i < n; ++i)
+        {
+            if (i > 0)
+                std::cout << ", ";
+            std::cout << it[i];
+        }
+        std::cout << std::endl;
+    }
+#endif // DUMP_CHECK_RESULTS
+
+    template <typename TValue, typename Iterator1, typename Iterator2, typename Iterator3, typename Size>
     void
-    check_values(Iterator1 host_keys, Iterator2 val_res, Size n)
+    check_values(Iterator1 host_keys, Iterator2 host_vals, Iterator3 val_res, Size n)
     {
-        //T keys[n1] = { 1, 2, 3, 4, 1, 1, 2, 2, 3, 3, 4, 4, 1, 1, 1, ...};
-        //T vals[n1] = { 1, 1, 1, 1, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 3, ...};
+        // https://docs.oneapi.io/versions/latest/onedpl/extension_api.html
+        // keys:   [ 0, 0, 0, 1, 1, 1 ]
+        // values: [ 1, 2, 3, 4, 5, 6 ]
+        // result: [ 1, 1 + 2 = 3, 1 + 2 + 3 = 6, 4, 4 + 5 = 9, 4 + 5 + 6 = 15 ]
+
+#ifdef DUMP_CHECK_RESULTS
+        std::cout << "check_values(n = " << n << ") : " << std::endl;
+        display_param("keys:   ", host_keys, n);
+        display_param("values: ", host_vals, n);
+        display_param("result: ", val_res,   n);
+#endif // DUMP_CHECK_RESULTS
+
+        if (n < 1)
+            return;
+
+        if (n == 1)
+        {
+            EXPECT_TRUE(host_vals[0] == val_res[0], "wrong effect from exclusive_scan_by_segment");
+            return;
+        }
 
-        int segment_length = 1;
-        auto expected_segment_sum = 1;
-        auto current_key = host_keys[0];
-        auto current_sum = 0;
-        for (int i = 0; i != n; ++i)
+        int segment_start_idx = 0;
```

Comment:
We have to a difference type of iterator  type instead of `int` (here and everywhere in the relevant places).

Reply:
Fixed